File: frontend/js/domain.js
import { apiPost, setLoading } from "/js/api.js";
import { clearElement, createProviderCard, createReportButton, createResultCard, createRiskCard, showError } from "/js/render.js";

const form = document.querySelector("#target-form");
const feedback = document.querySelector("#feedback");
const results = document.querySelector("#results");
const submitButton = form.querySelector("button[type='submit']");

function formatRecord(record) {
  if (typeof record === "string") {
    return record;
  }
  return Object.entries(record)
    .map(([key, value]) => `${key}: ${value}`)
    .join(" · ");
}

function dnsEntries(dns) {
  return Object.entries(dns.records).map(([recordType, recordSet]) => {
    const value = recordSet.status === "ok"
      ? recordSet.records.map(formatRecord).join(" | ")
      : recordSet.message || recordSet.status.replaceAll("_", " ");
    return [recordType, value];
  });
}

function renderDomain(data) {
  results.append(
    createResultCard(
      "Domain Summary",
      [
        ["Domain", data.domain],
        ["Unicode form", data.unicode_domain],
        ["Registered domain", data.registered_domain],
        ["Subdomain", data.subdomain],
        ["Punycode", data.punycode ? "Yes" : "No"],
      ],
      { badges: [{ label: "Domain" }] },
    ),
    createResultCard("DNS Records", dnsEntries(data.dns), { wide: true }),
  );
  renderProviders(data.threat_intelligence);
  renderRisk(data.risk_assessment);
}

function renderURL(data) {
  const query = data.query_parameters.length
    ? data.query_parameters.map(({ name, value }) => `${name}=${value}`).join(" · ")
    : "None";
  results.append(
    createResultCard(
      "URL Breakdown",
      [
        ["Refanged URL", data.refanged],
        ["Scheme", data.scheme],
        ["Host", data.host],
        ["Port", data.port],
        ["Path", data.path],
        ["Query parameters", query],
        ["Fragment", data.fragment || "None"],
        ["Registered domain", data.registered_domain],
        ["Subdomain", data.subdomain],
        ["HTTPS", data.https ? "Yes" : "No"],
      ],
      { badges: [{ label: "URL" }, { label: data.https ? "HTTPS" : "HTTP", muted: !data.https }] },
    ),
    createResultCard(
      "Suspicious Pattern Flags",
      data.flags.length
        ? data.flags.map((flag) => [flag.code.replaceAll("_", " "), flag.description])
        : [["Result", "No configured suspicious URL patterns matched."]],
    ),
  );
  if (data.dns) {
    results.append(createResultCard("Host DNS Records", dnsEntries(data.dns), { wide: true }));
  }
  results.append(
    createResultCard(
      "Disabled in v1",
      [
        ["TLS inspection", data.disabled_features.tls],
        ["Redirect following", data.disabled_features.redirect_chain],
      ],
      { wide: true, badges: [{ label: "SSRF protection", muted: true }] },
    ),
  );
  renderProviders(data.threat_intelligence);
  renderRisk(data.risk_assessment);
}

function renderProviders(providers) {
  for (const provider of Object.values(providers || {})) {
    results.append(createProviderCard(provider));
  }
}

function renderRisk(risk) {
  if (risk) {
    results.prepend(createRiskCard(risk));
  }
}

form.addEventListener("submit", async (event) => {
  event.preventDefault();
  clearElement(feedback);
  clearElement(results);
  setLoading(submitButton, true, "Analyzing…");

  try {
    const formData = new FormData(form);
    const analysis = await apiPost("/api/analyze/target", {
      target: formData.get("target"),
    });
    if (analysis.target_type === "url") {
      renderURL(analysis.result);
    } else {
      renderDomain(analysis.result);
    }
    results.append(createReportButton(analysis.investigation_id));
  } catch (error) {
    showError(feedback, error.message);
  } finally {
    setLoading(submitButton, false);
  }
});

File: backend/app/schemas/domain.py
from pydantic import BaseModel, ConfigDict, Field

from app.schemas.dns import DNSAnalysisResponse
from app.schemas.ip import IPAnalysisResponse
from app.schemas.threat import ProviderResultResponse
from app.schemas.risk import RiskAssessmentResponse


class DomainRequest(BaseModel):
    target: str = Field(
        min_length=1,
        max_length=256,
        examples=["example.com", "münich.example"],
    )


class URLRequest(BaseModel):
    url: str = Field(
        min_length=1,
        max_length=4096,
        examples=["https://example.com/login?source=email"],
    )


class TargetRequest(BaseModel):
    target: str = Field(
        min_length=1,
        max_length=4096,
        examples=["8.8.8.8", "example.com", "https://example.com/login"],
    )


class DomainAnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    input: str
    domain: str
    unicode_domain: str
    registered_domain: str
    subdomain: str | None
    punycode: bool
    dns: DNSAnalysisResponse
    investigation_id: int | None = None
    threat_intelligence: dict[str, ProviderResultResponse] = Field(default_factory=dict)
    risk_assessment: RiskAssessmentResponse | None = None


class URLFlagResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    description: str


class URLAnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    input: str
    refanged: str
    scheme: str
    host: str
    port: int
    explicit_port: bool
    path: str
    query_parameters: list[dict[str, str]]
    fragment: str
    userinfo: str | None
    registered_domain: str | None
    subdomain: str | None
    host_is_ip: bool
    https: bool
    flags: list[URLFlagResponse]
    dns: DNSAnalysisResponse | None
    tls: None
    redirect_chain: None
    disabled_features: dict[str, str]
    investigation_id: int | None = None
    threat_intelligence: dict[str, ProviderResultResponse] = Field(default_factory=dict)
    risk_assessment: RiskAssessmentResponse | None = None


class TargetAnalysisResponse(BaseModel):
    target_type: str
    result: IPAnalysisResponse | DomainAnalysisResponse | URLAnalysisResponse
    investigation_id: int
